feat: add pass headers guide (#6960)
vuestorefront/vue-storefront: docs/content/5.integrations/2.guide/2.adding-headers-to-request.md

# Pass Custom Headers in Requests

> In this tutorial, we'll demonstrate how to set up an interceptor to pass headers when making requests using the axios client inside the Vue Storefront SDK. This is useful when you need to append custom headers or cookies to your API calls.

Reviewer: Would be good to start such guide with some theoretical use-case scenario

After passing the headers to the axios client, we'll also show you how to pass them to external services using the API Client.

::danger

> Do not use createSharedComposable if creating a shared composable that uses the interceptor. This will cause the interceptor to only be created once, which will result in the headers being shared across instances.

Reviewer: It's a bit out of context here. A newcomer might have no idea what `createSharedComposable` is and why would he/she even has to use it. Perhaps this danger fits elsewhere?

Author: I agree it's no longer necessary. We have since removed this composable from our projects.

::

> ## Prerequisites:

Reviewer: Do we need to mention it if we're using Nuxt boilerplate that has SDK?

Ensure you have the Vue Storefront SDK set up in your project.

## Notes
::info

> This guide uses the integrations boilerplate with the nuxt option. The same basic principles apply to other frameworks, but the implementation may vary.

Reviewer: ```suggestion
This guide uses the [Nuxt Storefront Boilerplate](https://github.com/vuestorefront/storefront-nuxt3-boilerplate). for demonstration purposes. The same basic principles apply to other frameworks using the Vue Storefront stack, but the implementation may slightly vary.
```

Author: The guide does not use the nuxt 3 boilerplate. It uses the integrations boilerplate with the nuxt option.

::

> All SDK modules export a client instance of axios. This is the axios instance that is used to make requests. You can access this instance by importing the `client` from the SDK.

Reviewer: ```suggestion
All SDK modules export a client instance of [axios](https://axios-http.com/docs/intro), a popular HTTP Client. This is the axios instance that is used to make requests. You can access it by importing the `client` from the SDK.
```

> All SDK modules export a client instance of axios. This is the axios instance that is used to make requests. You can access this instance by importing the `client` from the SDK.

Reviewer: Please give a short example of how to access it and try to avoid describing code without showing it

> ## Steps

Reviewer: Would be great to have some more in-depth explanations of what is happening here and why. Since implementations can vary it's important for end-user to understand WHY these things should happen even more than WHAT should happen. Then they can adjust it to their projects.

::steps
#step-1
**Import Required Modules**

```javascript
import { initSDK, buildModule } from '@vue-storefront/sdk';
import { client, boilerplateModule, BoilerplateModuleType } from '../../packages/sdk/src';
import useHeaders from 'path-to-useHeaders-file'; // Replace with the correct path to the `useHeaders` composable.
```

#step-2
**Define the SDK Configuration**
```javascript
const sdkConfig = {
  boilerplate: buildModule<BoilerplateModuleType>(boilerplateModule, {
    apiUrl: 'http://localhost:8181/boilerplate',
  }),
};
```
#step-3
**Fetch Headers with the useHeaders Composable**

Use the `useHeaders` composable to fetch the necessary headers. This composable manages both server-side and client-side headers for you.

```javascript
const { addHeadersToState, headerData } = useHeaders();
addHeadersToState();
```

#step-4
**Setting Up Axios Interceptor**

> Before making requests using axios, establish an interceptor to automatically append the desired headers.

Reviewer 2: ```suggestion
Before making requests using axios, establish an interceptor to automatically append the desired headers.
```

Author: Did you intend to suggest a change?
<img width="809" alt="image" src="https://github.com/vuestorefront/vue-storefront/assets/45824492/898d7b91-68e7-41a5-8681-67642fbd8f75">

  const sdkConfig = {
    boilerplate: buildModule<BoilerplateModuleType>(boilerplateModule, {
      apiUrl: 'http://localhost:8181/boilerplate',
    }),
  };

  const {addHeadersToState, headerData} = useHeaders()

  addHeadersToState()

  // If an interceptor is already set, eject it to remove it
  // this prevents headers being shared across instances
  if (interceptorId !== null) {
    client.interceptors.request.eject(interceptorId);
  }

  interceptorId = client.interceptors.request.use(
    (config) => {
      if (!config.headers) {
        config.headers = {};
      }

      // here you can set any headers you want
      config.headers['amazing-header'] = 'coolest header ever';
      config.headers['dev-mode'] = headerData.value?.isDevMode ?? 'false'

      return config;
    },
    (error) => {
      // Do something with request error
      return Promise.reject(error);
    }
  );

  return initSDK<typeof sdkConfig>(sdkConfig);
};

```
::tip
This guide uses a `useHeaders` composable. This composable manages both server-side and client-side headers for you. This is just an example, you can use any method you want to fetch the headers.
::

#step-6

> **Create a Composable to handle headers**

Reviewer 2: ```suggestion
**Create a Composable to Handle Headers**
```
consistent capitalization with the other sections
